src/libopensc/card-piv.c

```diff
@@ -3657,88 +3657,138 @@ static int get_keylen(unsigned int alg_id, size_t *size)
 static int piv_get_key(sc_card_t *card, unsigned int alg_id, u8 **key, size_t *len)
 {
 
-	int r;
+	int r, r2;
 	size_t fsize;
 	FILE *f = NULL;
 	char * keyfilename = NULL;
 	size_t expected_keylen;
 	size_t keylen, readlen;
 	u8 * keybuf = NULL;
 	u8 * tkey = NULL;
+	u8 rbuf[255];
+	u8 sbuf[5];
+	u8 *p;
+	const u8 *body = NULL;
+	size_t body_len;
 
 	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
 
-	keyfilename = (char *)getenv("PIV_EXT_AUTH_KEY");
-
-	if (keyfilename == NULL) {
-		sc_log(card->ctx,
-			"Unable to get PIV_EXT_AUTH_KEY=(null) for general_external_authenticate");
-		r = SC_ERROR_FILE_NOT_FOUND;
-		goto err;
-	}
-
 	r = get_keylen(alg_id, &expected_keylen);
 	if(r) {
 		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "Invalid cipher selector, none found for:  %02x", alg_id);
 		r = SC_ERROR_INVALID_ARGUMENTS;
 		goto err;
 	}
 
-	f = fopen(keyfilename, "rb");
-	if (!f) {
-		sc_log(card->ctx, " Unable to load key from file\n");
-		r = SC_ERROR_FILE_NOT_FOUND;
-		goto err;
-	}
-
-	if (0 > fseek(f, 0L, SEEK_END))
-		r = SC_ERROR_INTERNAL;
-	fsize = ftell(f);
-	if (0 > (long) fsize)
-		r = SC_ERROR_INTERNAL;
-	if (0 > fseek(f, 0L, SEEK_SET))
-		r = SC_ERROR_INTERNAL;
-	if(r) {
-		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "Could not read %s\n", keyfilename);
-		goto err;
-	}
-
-	keybuf = malloc(fsize+1); /* if not binary, need null to make it a string */
-	if (!keybuf) {
-		sc_log(card->ctx, " Unable to allocate key memory");
-		r = SC_ERROR_OUT_OF_MEMORY;
-		goto err;
-	}
-	keybuf[fsize] = 0x00;    /* in case it is text need null */
-
-	if ((readlen = fread(keybuf, 1, fsize, f)) != fsize) {
-		sc_log(card->ctx, " Unable to read key\n");
-		r = SC_ERROR_WRONG_LENGTH;
-		goto err;
-	}
-	keybuf[readlen] = '\0';
-
 	tkey = malloc(expected_keylen);
 	if (!tkey) {
 		sc_log(card->ctx, " Unable to allocate key memory");
 		r = SC_ERROR_OUT_OF_MEMORY;
 		goto err;
 	}
 
-	if (fsize == expected_keylen) { /* it must be binary */
-		memcpy(tkey, keybuf, expected_keylen);
+	p = sbuf;
+	*p++ = 0x5C;
+	*p++ = 0x03;
+	*p++ = 0x2F;
+	*p++ = 0x41;
+	*p++ = 0x4B;
+
+	r2 = piv_general_io(card, 0xCB, 0x3F, 0x00, sbuf, p - sbuf, rbuf, sizeof rbuf);
```

Comment:
Could you add a comment that explains what you are doing here?

> Fetch Status Object to...

A description of the status object structure would also help.